tap_mssql/sync_strategies/log_based.py

```diff
@@ -1,396 +1,405 @@
+                            " for extract table %s, and the maximum lsn is %s."
+                        ),
+                        state_last_lsn,
+                        lsn_from,
+                        lsn_to,
+                    )
+                    if lsn_to == state_last_lsn:
+                        LOGGER.info(
+                            (
+                                "The last lsn processed as per the state file is equal to the max"
+                                " lsn available - no changes expected - state lsn will not be incremented"
+                            ),
+                        )
+                        from_lsn_expression = "{}".format(py_bin_to_mssql(state_last_lsn))
+                    else:
+                        from_lsn_expression = (
+                            (
+                                "sys.fn_cdc_increment_lsn({})"
+                            ).format(py_bin_to_mssql(state_last_lsn))
+                        )
+                else:
+                    raise Exception(
+                        (
+                            "Error {}.{}: CDC changes have expired, the minimum lsn is {}, the last"
+                            " processed lsn is {}. Recommend a full load as there may be missing data."
+                        ).format(escaped_schema_name, escape_table_name, lsn_from, state_last_lsn)
+                    )
+
+                select_sql = """DECLARE @from_lsn binary (10), @to_lsn binary (10)
+
+                                SET @from_lsn = {}
+                                SET @to_lsn = {}
+
+                                SELECT {}
+                                    ,case __$operation
+                                        when 2 then 'I'
+                                        when 4 then 'U'
+                                        when 1 then 'D'
+                                    end _sdc_operation_type
+                                    , sys.fn_cdc_map_lsn_to_time(__$start_lsn) _sdc_lsn_commit_timestamp
+                                    , case __$operation
+                                        when 1 then sys.fn_cdc_map_lsn_to_time(__$start_lsn)
+                                        else null
+                                        end _sdc_lsn_deleted_at
+                                    , __$start_lsn _sdc_lsn_value
+                                    , __$seqval _sdc_lsn_seq_value
+                                    , __$operation _sdc_lsn_operation
+                                FROM cdc."fn_cdc_get_all_changes_{}"(@from_lsn, @to_lsn, 'all')
+                                ORDER BY __$start_lsn, __$seqval, __$operation
+                                ;""".format(
+                    from_lsn_expression,
+                    py_bin_to_mssql(lsn_to),
+                    ",".join(escaped_columns),
+                    cdc_table,
+                )
+
```

Comment:
Providing a consistent approach for escaping SQL Server objects might be preferable - my suggestion here adds a variable containing the cdc function name which is then escaped, then passed into the SQL.

```suggestion
                escaped_cdc_function = common.escape("fn_cdc_get_all_changes_" + cdc_table)

                select_sql = """DECLARE @from_lsn binary (10), @to_lsn binary (10)

                                SET @from_lsn = {}
                                SET @to_lsn = {}

                                SELECT {}
                                    ,case __$operation
                                        when 2 then 'I'
                                        when 4 then 'U'
                                        when 1 then 'D'
                                    end _sdc_operation_type
                                    , sys.fn_cdc_map_lsn_to_time(__$start_lsn) _sdc_lsn_commit_timestamp
                                    , case __$operation
                                        when 1 then sys.fn_cdc_map_lsn_to_time(__$start_lsn)
                                        else null
                                        end _sdc_lsn_deleted_at
                                    , __$start_lsn _sdc_lsn_value
                                    , __$seqval _sdc_lsn_seq_value
                                    , __$operation _sdc_lsn_operation
                                FROM cdc.{}(@from_lsn, @to_lsn, 'all')
                                ORDER BY __$start_lsn, __$seqval, __$operation
                                ;""".format(
                    from_lsn_expression,
                    py_bin_to_mssql(lsn_to),
                    ",".join(escaped_columns),
                    escaped_cdc_function,
                )

```